File: Arrays & Strings Javascript/StringCompression.js
// String compression

function stringCompression(string) {
    // string must be letters only
    if (string.match(/[^a-z]/i)) {
        return 'String must be letters only';
    }

    let compressed = '';
    let count = 0;
    for (let i = 0; i < string.length; i++) {
        count++;
        if (string[i] !== string[i + 1]) {
            compressed += string[i] + count;
            count = 0;
        }
    }
    return compressed.length < string.length ? compressed : string;
}

console.log(stringCompression('aabcc23cccaaa'));
console.log(stringCompression('abc'));
console.log(stringCompression('aabbcc'));

File: Arrays & Strings Python/stringCompression.py
# String Compression: Implement a method to perform basic string compression using the counts
# of repeated characters. For example, the string aabcccccaaa would become a2blc5a3. If the
# "compressed" string would not become smaller than the original string, your method should return
# the original string. You can assume the string has only uppercase and lowercase letters (a - z).

def stringCompression(string):
    # check if there are no digits in the string
    if any(char.isdigit() for char in string):
        return "no digits allowed in string"

    newString = ''
    count = 0

    for i in range(len(string)):
        count += 1

        if i + 1 >= len(string) or string[i] != string[i + 1]:
            newString += string[i] + str(count)
            count = 0

    return newString if len(newString) < len(string) else string


print(stringCompression("aaaabbbbccyyycc"))
